### PR TITLE
Move jobrunner to using a docker image with autodeployment.

### PR DESCRIPTION
**Pull request by bloodearnest**
This is patterned on how we run airlock:
 - uses docker compose to run and manage restarts and logs
 - a systemd timer/service to regularly pull down new images

We get to remove the code/lib dirs, and some of the awkward scripting.

Jobrunner has a lot of management commands, but those are mostly
abstracted behind just, and I have added any that were missing to the
justfile.

Deploying this will require some post install manuall cleanup of
code/lib dirs.

It was difficult to break up into small commits, at its a big refactor
that touches a lot of related things.

### CODE REVIEW

File: playbook.md [hunk 1]
```diff
@@ -19,17 +19,18 @@ the opensafely user with:
...
-The jobrunner is installed in `/home/opensafely/jobrunner`
 
-    /home/opensafely/jobrunner/code     # the checkout of jobrunner currently running
-    /home/opensafely/jobrunner/lib      # dependencies for jobrunner, added via PYTHONPATH
-    /home/opensafely/config  # environment configuration
-    /home/opensafely/secret   # any secret files (e.g. x509 client certificates for emis)
+/home/opensafely/config  # environment configuration
+/home/opensafely/secret   # any secret files (e.g. x509 client certificates for emis)
+/home/opensafely/jobrunner  # jobrunner service workdir and configuration
+/home/opensafely/airlock    # airlcok service workdir and configuration
```
Comment (madwort) [inline playbook.md:28]:
```suggestion
/home/opensafely/airlock    # airlock service workdir and configuration
```

File: services/jobrunner/install.sh [hunk 2]
```diff
@@ -15,45 +14,34 @@ umask 027
...
-chown -R opensafely:opensafely "$HIGH_PRIVACY_STORAGE_BASE"
-chown -R "opensafely:$REVIEWERS_GROUP" "$MEDIUM_PRIVACY_STORAGE_BASE"
 
+chown -R opensafely:opensafely "$HIGH_PRIVACY_STORAGE_BASE" "$MEDIUM_PRIVACY_STORAGE_BASE"
 
 # create initial db if not present
-export PYTHONPATH=$TARGET_DIR/lib:$TARGET_DIR/code
-workdir=$(python3 -c "from jobrunner import config; print(config.WORKDIR)")
-test -f "$workdir/db.sqlite" || python3 -m jobrunner.cli.migrate
+test -f "$DIR/workdir/db.sqlite" || just -f $DIR/justfile migrate
```
Comment (madwort) [inline services/jobrunner/install.sh:38]: will this leave `db.sqlite` (incorrectly) owned by root?

File: services/jobrunner/install.sh [hunk 3]
```diff
@@ -15,45 +14,34 @@ umask 027
...
 
-# ensure file ownership and permissions
-chown -R opensafely:opensafely $TARGET_DIR
-chown -R opensafely:opensafely $HOME_DIR
```
Comment (madwort) [inline services/jobrunner/install.sh:44]: I would be tempted to keep these (with corrections for the changed envvar names obvs)

File: services/jobrunner/install.sh (hunk 2 above)
Comment (madwort) [reply services/jobrunner/install.sh:38]: hmm, maybe not as the justfile should be run by the `opensafely` user?

Comment (madwort) [reply services/jobrunner/install.sh:38]: in any case, deployment scripts shouldn't care which user ran them

File: services/jobrunner/justfile [hunk 4]
```diff
@@ -3,50 +3,51 @@
...
+
+# Update jobrunner and restart
+deploy: && restart
+    echo "Pulling image"
+    docker compose pull --quiet jobrunner
+    
 # run jobrunner db migrations
 migrate:
-    python3 -m jobrunner.cli.migrate
+    docker compose run jobrunner python3 -m jobrunner.cli.migrate
```
Comment (madwort) [inline services/jobrunner/justfile:30]: this is ok, although maybe we could use [jobrunner's justfile](https://github.com/opensafely-core/job-runner/blob/main/justfile#L117-L118) and do something like this?

```
docker compose run jobrunner just migrate
```

File: services/jobrunner/justfile [hunk 5]
```diff
@@ -3,50 +3,51 @@
...
 
 # destroy containers & volumes for all running jobs
 prepare-for-reboot: stop
-    python3 -m jobrunner.cli.prepare_for_reboot
+    docker compose run jobrunner python3 -m jobrunner.cli.prepare_for_reboot
```
Comment (madwort) [inline services/jobrunner/justfile:34]: (same)

File: services/jobrunner/justfile [hunk 6]
```diff
@@ -3,50 +3,51 @@
...
 
 # stop accepting new jobs
 pause:
-    python3 -m jobrunner.cli.flags set paused=true
+    docker compose run jobrunner python3 -m jobrunner.cli.flags set paused=true
```
Comment (madwort) [inline services/jobrunner/justfile:38]: (same)

File: services/jobrunner/justfile [hunk 7]
```diff
@@ -3,50 +3,51 @@
...
 
 # start accepting new jobs
 unpause:
-    python3 -m jobrunner.cli.flags set paused=
+    docker compose run jobrunner python3 -m jobrunner.cli.flags set paused=
```
Comment (madwort) [inline services/jobrunner/justfile:42]: (same)

File: services/jobrunner/justfile [hunk 8]
```diff
@@ -3,50 +3,51 @@
...
 
-# show jobrunner logs (limited to 1000 lines) 
-logs lines="1000":
-    journalctl -xe -u jobrunner -n{{ lines }}
+# show jobrunner logs
```
Comment (madwort) [inline services/jobrunner/justfile:44]: helpful or unnecessary?

```suggestion
# show jobrunner logs (using `docker compose logs`)
```

File: services/jobrunner/justfile [hunk 9]
```diff
@@ -58,12 +59,31 @@ jobs-stats:
...
+
+# update docker image
+update-docker-image image:
+    #!/bin/bash
+    set -euo pipefail
+    # pull from proxy 
+    docker pull docker-proxy.opensafely.org/opensafely-core/{{ image }}
+    # tag as if we'd pulled direct from ghcr.io 
+    docker tag docker-proxy.opensafely.org/opensafely-core/{{ image }} ghcr.io/opensafely-core/{{ image }}
+    docker image prune --force
```
Comment (madwort) [inline services/jobrunner/justfile:84]: aside: when I was updating the ehrql docker image yesterday, it pruned some airlock images - which was unexpected! but not /incorrect/ as such...

**Review by madwort — APPROVED** [review]
looks good, hopefully it survives contact with reality.
Maybe we could go further & push some management tasks into job-runner's own justfile, but that doesn't have to be in this PR (& maybe it's better if it's not!)

File: services/jobrunner/install.sh (hunk 2 above)
Comment (bloodearnest) [reply services/jobrunner/install.sh:38]: It uses docker compose to run the migrate command inside the container, and compose is configured to always run as opensafely user, regardless of who invokes docker compose.

If db.sqlite was owned by root, I would expect the test to fail.

Comment (madwort) [reply services/jobrunner/install.sh:38]: ah yes of course, that's fine then

**Pull request merged**